cli/src/cmd/forge/script/broadcast.rs

```diff
@@ -279,43 +279,105 @@ impl ScriptArgs {
-            match new_txes.front().unwrap().typed_tx() {
-                TypedTransaction::Legacy(_) | TypedTransaction::Eip2930(_) => {
-                    provider.get_gas_price().await?
+        if !self.skip_simulation {
+            // We don't store it in the transactions, since we want the most updated value. Right
+            // before broadcasting.
+            let per_gas = if let Some(gas_price) = self.with_gas_price {
+                gas_price
+            } else {
+                match new_txes.front().unwrap().typed_tx() {
+                    TypedTransaction::Legacy(_) | TypedTransaction::Eip2930(_) => {
+                        provider.get_gas_price().await?
+                    }
+                    TypedTransaction::Eip1559(_) => provider.estimate_eip1559_fees(None).await?.0,
                 }
-                TypedTransaction::Eip1559(_) => provider.estimate_eip1559_fees(None).await?.0,
-            }
-        };
+            };
 
-        println!("\n==========================");
-        println!("\nEstimated total gas used for script: {}", total_gas);
-        println!(
-            "\nEstimated amount required: {} ETH",
-            format_units(total_gas.saturating_mul(per_gas), 18)
-                .unwrap_or_else(|_| "[Could not calculate]".to_string())
-                .trim_end_matches('0')
-        );
-        println!("\n==========================");
+            println!("\n==========================");
+            println!("\nEstimated total gas used for script: {}", total_gas);
+            println!(
+                "\nEstimated amount required: {} ETH",
+                format_units(total_gas.saturating_mul(per_gas), 18)
+                    .unwrap_or_else(|_| "[Could not calculate]".to_string())
+                    .trim_end_matches('0')
+            );
+            println!("\n==========================");
+        }
         Ok(new_txes)
     }
+    /// Uses the signer to submit a transaction to the network. If it fails, it tries to retrieve
+    /// the transaction hash that can be used on a later run with `--resume`.
+    async fn broadcast<T, U>(
+        &self,
+        signer: &SignerMiddleware<T, U>,
+        mut legacy_or_1559: TypedTransaction,
+    ) -> Result<TxHash, BroadcastError>
+    where
+        T: Middleware,
+        U: Signer,
+    {
+        tracing::debug!("sending transaction: {:?}", legacy_or_1559);
+
+        // Chains which use `eth_estimateGas` are being sent sequentially and require their gas to
+        // be re-estimated right before broadcasting.
+        if has_different_gas_calc(signer.signer().chain_id()) || self.skip_simulation {
+            // if already set, some RPC endpoints might simply return the gas value that is already
+            // set in the request and omit the estimate altogether, so we remove it here
+            let _ = legacy_or_1559.gas_mut().take();
+
+            self.estimate_gas(&mut legacy_or_1559, signer.provider()).await?;
```

Comment:
Can you add context to this error? `.wrap_err_with(|| format("Failed to estimate gas bla bla bla"))?`

Reply:
Good point. I added the context inside the `estimate_gas` method to have it consistent across calls:
https://github.com/foundry-rs/foundry/pull/2601/files#diff-2b2997f85d225c8c6fdd337f5382a5be0fbd7f254f4d7854161576400646e563R375